src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -147,6 +148,13 @@ public final void reset(Collection<? extends Whitelist> delegates) {
         }
     }
 
+    private void addWrapper(ProxyWhitelist pwl){
+        lock.writeLock().lock();
```

Comment:
Could we get deadlock now in some cases? `#reset` holds a lock for the outer ProxyWhitelist, and when this is called it will acquire the lock for the delegate ProxyWhitelist. Do we know that the locks will always be acquired in the same order? I would be surprised if there are cyclic references between instances of ProxyWhitelist, so my guess is that it should be fine, but it would be good for someone to confirm.

Reply:
Cyclic references shouldn't be possible, but this makes me nervous too -- although for different reasons: I wonder what happens if we have two different `reset` operations running in parallel, starting from different ProxyWhitelists.  

Note also that read and write locks are independent - you can't "upgrade" a readlock to a writelock.

Reply:
AFAICT we actually ought to obtain one of the delegate's locks  when reading its fields, meaning this whole block theoretically should be synchronized (probably with a WriteLock for safety, since it also implies a readlock).

Reply:
Theoretically the lock orders should be consistent, since we're climbing up the tree of wrapping whitelists, but I worry about the fact that we also call reset on wrapper ProxyWhitelists further down (and have read calls coming in as well).   WDYT @jglick ?

Reply:
@dwnusbaum it makes me worry too, because I am not sure this is impossible too. I will think about writing at least some test for this, but honestly I wanted to write it for whole reset method and it is not such easy task. I was able to hit my issue with test but not in 100%, usually I hit it in 2 cases from 5. Increasing the number of loops did not make it more successful. The whole method made it less probably with increasing loops, since growing numbers of elements in map (since the wrappers are only added but never removed) increases time of execution somewhere else then where the issues is.

Reply:
> WDYT @jglick ?

It would take me a while to digest this and remember exactly how this code works. Any change here is pretty risky.

Reply:
@jglick @dwnusbaum 
Is this even needed?   We're already taking the write lock at the top of the calling method and there are not other places where this private field is accessed.

Reply:
@bitwiseman The premise of the PR is that without a lock here, @lvotypko ran into a situation where `wrappers` was updated concurrently which caused an infinite loop. If we don't add a lock here, then this PR doesn't do anything. My best guess as to what happened is that two distinct whitelists delegated to the same whitelist, and reset was somehow called on them concurrently, but without a way to reproduce the issue it's hard to say.